### PR TITLE
parser: Recover from `elif` as `else if`

### LINKED ISSUE #110169: Better handling of `elif -> else if` syntax error
### Code

```Rust
if 1 > 2 {
  println!("Hello.");
} elif 2 > 1 {
  println!("Bye.");
}
```

### Current output

```Shell
Compiling playground v0.0.1 (/playground)
error: expected one of `!`, `.`, `::`, `;`, `?`, `{`, `}`, or an operator, found `2`
 --> src/main.rs:3:8
  |
3 | } elif 2 > 1 {
  |        ^ expected one of 8 possible tokens

error: could not compile `playground` (bin "playground") due to previous error
```

### Desired output

```Shell
Compiling playground v0.0.1 (/playground)
error: expected one of `!`, `.`, `::`, `;`, `?`, `{`, `}`, or an operator, found `2`
 --> src/main.rs:3:8
  |
3 | } elif 2 > 1 {
  |   ---- ^ expected one of 8 possible tokens
  |   |
  |   help: write `else if` instead of `elif`

error: could not compile `playground` (bin "playground") due to previous error
```

### Rationale and extra context

Writing `elif` instead of `else if` would be a very common error for coders coming from Python (or other some languages).

### Other cases

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by clubby789**
Fixes #110169

This also moves all `test/ui/parser/recover-*` to `test/ui/parser/recover` to fix a tidy error.
2 commits included so the diff in diagnostics can be seen (3bed1daf7a8dc98f5f356c352e9f68b9d36f01df), but I can squash this into one if that's preferred

### CODE REVIEW

File: compiler/rustc_parse/src/parser/expr.rs [hunk 1]
```diff
@@ -2332,7 +2332,15 @@ impl<'a> Parser<'a> {
             self.error_on_if_block_attrs(lo, false, block.span, attrs);
             block
         };
-        let els = if self.eat_keyword(kw::Else) { Some(self.parse_expr_else()?) } else { None };
+        let els = if self.eat_keyword(kw::Else) {
+            Some(self.parse_expr_else(false)?)
+        } else if self.may_recover() && self.token.is_keyword(sym::elif) {
```
Comment (lukas-code) [inline compiler/rustc_parse/src/parser/expr.rs:2337]: Doesn't this break code like this?

```rust
fn main() {
    let elif = ();
    if true {} elif
}
```

Comment (clubby789) [reply compiler/rustc_parse/src/parser/expr.rs:2337]: Hmm, good point. I'll see if I can come up with a better solution

**Comment by clubby789** [comment]
Hmm, I guess I could pass down `recovering_elif` all the way down but that seems a bit messy. Is there any way to delay a diagnostic to only be emitted if an error is, i.e. the opposite of `delay_span_bug`?

**Comment by compiler-errors** [comment]
> Is there any way to delay a diagnostic to only be emitted if an error is

No, not as far as I know.

**Comment by clubby789** [comment]
I think I'll close this for now, unless I can think of a good way of doing this :/

**Pull request closed without merging**